src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php

```diff
@@ -0,0 +1,183 @@
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Category\NameBuilder;
+
+use Cache;
+use PrestaShop\PrestaShop\Adapter\Category\Repository\CategoryRepository;
+use PrestaShop\PrestaShop\Core\Domain\Category\ValueObject\CategoryId;
+use PrestaShop\PrestaShop\Core\Domain\Language\ValueObject\LanguageId;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopId;
+
+/**
+ * Builds category display name to avoid confusing categories when they have identical names in multiple levels.
+ */
+class CategoryDisplayNameBuilder
+{
+    /**
+     * @var CategoryRepository
+     */
+    private $categoryRepository;
+
+    /**
+     * @param CategoryRepository $categoryRepository
+     */
+    public function __construct(
+        CategoryRepository $categoryRepository
+    ) {
+        $this->categoryRepository = $categoryRepository;
+    }
+
+    /**
+     * @see buildDisplayName
+     *
+     * @param CategoryId $categoryId
+     * @param string $categoryName
+     * @param ShopId $shopId
+     * @param LanguageId $languageId
+     * @param bool $useCache
+     *
+     * @return string
+     */
+    public function build(
+        CategoryId $categoryId,
+        string $categoryName,
+        ShopId $shopId,
+        LanguageId $languageId,
+        bool $useCache = true
+    ): string {
+        return $this->buildDisplayName($categoryId, $categoryName, $shopId, $languageId, null, $useCache);
+    }
+
+    /**
+     * @see buildDisplayName
+     *
+     * In some cases we already have breadcrumbs before calling this method, so we allow providing them here for optimization purposes
+     *
+     * @param CategoryId $categoryId
+     * @param string $categoryName
+     * @param ShopId $shopId
+     * @param LanguageId $languageId
+     * @param string[] $breadcrumbParts
+     * @param bool $useCache
+     *
+     * @return string
+     */
+    public function buildWithBreadcrumbs(
+        CategoryId $categoryId,
```

Comment:
```suggestion
```

Reply:
Seems like `$categoryId` is only used by `CategoryRepository::getBreadcrumbParts` so it's not useful for this method which already has breadcrumb parts provided, no need to keep it in this public method